### PR TITLE
Generalize font-alias mechanism

### LINKED ISSUE #73: Generalize the font-alias mechanism
Currently, "Arial" is defined in the code as an alias for "helvetica":
- [in `add_font`](https://github.com/PyFPDF/fpdf2/blob/2.2.0/fpdf/fpdf.py#L581)
- [in `set_font`](https://github.com/PyFPDF/fpdf2/blob/2.2.0/fpdf/fpdf.py#L709)

This could be generalized according to the table of alternatives from appendix H, section 5.5.1 of the [PDF 1.7 spec](https://ghostscript.com/~robin/pdf_reference17.pdf):

Standard name           | Alternative
------------------------|-
Courier                 | CourierNew
Courier − Oblique       | CourierNew,Italic
Courier − Bold          | CourierNew,Bold
Courier − BoldOblique   | CourierNew,BoldItalic
Helvetica               | Arial
Helvetica − Oblique     | Arial,Italic
Helvetica − Bold        | Arial,Bold
Helvetica − BoldOblique | Arial,BoldItalic
Times − Roman           | TimesNewRoman
Times − Italic          | TimesNewRoman,Italic
Times − Bold            | TimesNewRoman,Bold
Times − BoldItalic      | TimesNewRoman,BoldItalic

### PR DESCRIPTION
**Pull request by fbernhart**
Closes #73 

### Changes to fpdf.py

- I've introduced a new attribute to the `FPDF` class: `self.font_aliases`, a dict which will keep track of the different font aliases, e.g. Helvetica as alternative for Arial
- I've changed the default value of `uni` from `False` to `True` in the `add_font` method. According to the [fpdf documentation](https://pyfpdf.github.io/fpdf2/reference/add_font.html#parameters):
  > Calling this method with uni=False is discouraged as legacy font support is complex and deprecated.

  So I thought it would be a good idea to change the default behaviour to `True`.
  
- I've more or less completely reworked the `set_font` method and tried to simplify the code. "Arial", "CourierNew" and "TimesNewRoman" are now replaced by their aliases (if they haven't been added before). I've as well initialized `self.font_size` and `self.current_font` in the `__init__` method.
- `add_font` and `set_font` are now always working with lowercase font names internally. 

### Changes to test_fonts.py

- I've added some tests to test the new "font aliases" feature.
- I've as well added some additional tests to test the `set_font` function for different inputs.

### CODE REVIEW

File: fpdf/fpdf.py [hunk 1]
```diff
@@ -148,6 +148,7 @@ def __init__(self, orientation="P", unit="mm", format="A4"):
         self.links = {}  # array of internal links
         self.in_footer = 0  # flag set when processing footer
         self.lasth = 0  # height of last cell printed
+        self.current_font = {}  # current font
```
Comment (Lucas-C) [inline fpdf/fpdf.py:151]: Seems unused

**Comment by Lucas-C** [comment]
I think making `True` the default value for `uni` in `add_font` is a good idea,
but this will have bad consequences on users code relying on the default value:
they will have bad surprises when upgrading to the latest `fpdf2` version!

If you really really think to perform this change, I recommend the following steps:
1. do this in a dedicated PR
2. consider renaming the parameter instead of just changing its default value.
This way, we could deprecate the `uni` parameter and display a warning to users still using it.
3. document this change in `CHANGELOG.md`
4. update all the documentation snippets to the correct usage of `uni` or the new parameter

Also, about this comment in this PR description:
> add_font and set_font are now always working with lowercase font names internally.

Do you imply they did not previously?
There is a test validating that behaviour with `set_font` though:
https://github.com/PyFPDF/fpdf2/blob/master/test/fonts/test_fonts.py#L45

**Comment by Lucas-C** [comment]
Just FYI, if you reference issue #73 in your commit message,
it will be auto-closed: https://github.blog/2013-05-14-closing-issues-via-pull-requests/

**Comment by Lucas-C** [comment]
You will have to rebase this PR I think.

Also, could you add a mention of this font-alias generic mechanism in `CHANGELOG.md` please?
And you can also add your name to `CONTRIBUTORS.md` 😉

**Comment by alexanderankin** [comment]
this looks good but i think that using an alias shouldn't issue a warning but using an invalid combination like bold with non-boldable fonts should throw an error (and be able to be configured not to). in an ideal world there would be a separate class (or two) for managing font data and rendering it to pdf format. I'm not sure how to push commits to this PR but I can post a diff here of what i would change.

Please dont change uni to False.

Re the default unicode version - [the original interface does not specify this argument](http://www.fpdf.org/en/doc/addfont.htm) so its thats an artifact of this library and now we have to figure out what to do with it. Except that they had a `MakeFont` thing, though, so they couldnt figure it out either. best thing is `add_font` and `add_uni_font` but short of that, lets not change our own api. I'm sure its somehow possible to do it without caching (with a good font library, but thats way beyond the scope)

**Comment by fbernhart** [comment]
Thanks for the feedback. 👍🏼 Right now I'm not having a lot of free time, but I'll get back to this PR as soon as possible.

**Comment by Lucas-C** [comment]
I agree about the log level: we could move this WARN message about alias into an INFO or DEBUG log message.

**Comment by alexp1917** [comment]
i started looking at this and this conflict is actually about 90 lines after merging into master locally ` 1 file changed, 65 insertions(+), 28 deletions(-)`. Its pretty late here but ill see if i can merge it, if neither of you guys gets to it first.

**Comment by alexp1917** [comment]
i started looking at this and this conflict is actually about 90 lines after merging into master locally ` 1 file changed, 65 insertions(+), 28 deletions(-)`. Its pretty late here but ill see if i can merge it, if none of you guys gets to it first.

**Comment by fbernhart** [comment]
@alexp1917 I already solved the conflicts yesterday locally. I assume on the weekend I'll have time to finish this PR.

File: fpdf/fpdf.py (hunk 1 above)
Comment (fbernhart) [reply fpdf/fpdf.py:169]: @Lucas-C It's used for example here:

`self.current_font = self.fonts[fontkey]`

and as well in many other places throughout the `fpdf.py` file. I just initialized it.

**Comment by fbernhart** [comment]
@Lucas-C I've now changed back the `uni` parameter back to `False` as default. If there is a need to change it, it should indeed happen in an extra PR. 👍🏼 
<br>
<br>

> Also, about this comment in this PR description:
>
> >add_font and set_font are now always working with lowercase font names internally.
>
>Do you imply they did not previously?

Yes. Before it was possible to add `Arial` as well as `arial` (or `ARIAL`, `aRiaL` etc.) as font. Now the name is normalized to lowercase.
<br>
<br>

> Just FYI, if you reference issue #73 in your commit message, it will be auto-closed https://github.blog/2013-05-14-closing-issues-via-pull-requests/

Yeah, I know. That was my intention.
<br>
<br>

> I agree about the log level: we could move this WARN message about alias into an INFO or DEBUG log message.

Hmm ... I haven't worked much with warnings/logging etc. yet. But according to the [Python documentation](https://docs.python.org/3/howto/logging.html):

>`warnings.warn()` in library code if the issue is avoidable and the client application should be modified to eliminate the warning

Which sounds very reasonable to, as setting for example Zapfdingbats with italic as style is avoidable. Compared to for example:

> `INFO`: Confirmation that things are working as expected.

**Comment by alexp1917** [comment]
i was able to comment this out and the tests passed:

```diff
diff --git a/fpdf/fpdf.py b/fpdf/fpdf.py
index 960f582..3d17eb0 100644
--- a/fpdf/fpdf.py
+++ b/fpdf/fpdf.py
@@ -217,7 +217,7 @@ class FPDF:
         else:
             raise FPDFException(f"Incorrect unit: {unit}")
 
-        self.font_size = self.font_size_pt / self.k  # current font size
+        # self.font_size = self.font_size_pt / self.k  # current font size
 
         # Page format
         self.dw_pt, self.dh_pt = get_page_format(format, self.k)
```

**Comment by alexp1917** [comment]
likewise, this should also be removed:

```diff
diff --git a/test/fonts/test_add_font.py b/test/fonts/test_add_font.py
index efac787..2480b95 100644
--- a/test/fonts/test_add_font.py
+++ b/test/fonts/test_add_font.py
@@ -77,7 +77,7 @@ def test_add_font_from_pkl(tmp_path):
     pdf = FPDF()
     font_file_path = HERE / "../end_to_end_legacy/charmap/Roboto-Regular.ttf"
     pdf.add_font("Roboto-Regular", fname=font_file_path, uni=True)
-    del pdf.fonts["roboto-regular"]  # required to be allow to re-add it
+    # del pdf.fonts["roboto-regular"]  # required to be allow to re-add it
     # re-using .pkl file generated by previous method call:
     pdf.add_font("Roboto-Regular")
     pdf.set_font("Roboto-Regular", size=64)

```

**Comment by alexp1917** [comment]
lastly substituting is misspelled (per google), should be one t:

```diff
diff --git a/fpdf/fpdf.py b/fpdf/fpdf.py
index 960f582..6964fc1 100644
--- a/fpdf/fpdf.py
+++ b/fpdf/fpdf.py
@@ -823,7 +823,7 @@ class FPDF:
 
         if family in self.font_aliases and family + style not in self.fonts:
             warnings.warn(
-                f"Substitutting font {family} by core font "
+                f"Substituting font {family} by core font "
                 f"{self.font_aliases[family]}"
             )
             family = self.font_aliases[family]
```

and with those minor adjustments, i would love to merge this.

**Comment by fbernhart** [comment]
@alexp1917 

> i was able to comment this out and the tests passed:
>
> `self.font_size = self.font_size_pt / self.k  # current font size`

I just wanted to initalize self.font_size in order to keep a good coding style. But if you want, I can still remove it.


> likewise, this should also be removed:
>
> `del pdf.fonts["roboto-regular"]  # required to be allow to re-add it`

I didn't write this test. I just adapted it slightly in order to work with my PR (lower case font names). So I don't wanna mess it up by deleting lines without fully understanding it.


> lastly substituting is misspelled in the test and implementation (per google), should be one t:

Fixed. Thanks 👍🏼

**Comment by alexp1917** [comment]
@fbernhart yes please remove it as the correct way would be to call `self.set_font()` from the `__init__` except that this method is (currently) required to be called afterwards anyway.

**Comment by fbernhart** [comment]
@alexp1917 Thanks. I've removed the initialisation.

**Review by alexanderankin — COMMENTED** [review]
I just got a whole bunch of emails for this, im not maintaining this library anymore so please request review from @alexp1917

**Comment by fbernhart** [comment]
@alexp1917 Thanks for merging 👍🏼

@alexanderankin I've never requested a review from you. 🤔

**Comment by Lucas-C** [comment]
Nice work on this @fbernhart !

**Pull request merged**